### ISSUE TITLE
Create an official repository for CI helpers (aka Actions)

### ISSUE THREAD

**Issue by umarcor** [godotengine/godot-proposals#1413]
**Describe the project you are working on:**

Not relevant. This request is usable for any project that needs to be built automatically and periodically.

**Describe the problem or limitation you are having in your project:**

While trying to setup CI for a Godot project in GitHub, using GitHub Actions, I found that 6 different projects are available in the marketplace: https://github.com/marketplace?type=actions&query=godot. On the one hand, none of them seems to be officially supported by the org. On the other hand, many of them feel redundant, but there is no discussion about the differences/motivations.

I did not check *helpers* for other CI services, but I'd expect the ecosystem to be either similar or inexistent.

**Describe the feature / enhancement and how it helps to overcome the problem or limitation:**

Regardless of the CI service, most users willing to export their projects/games/tools to multiple platforms are likely to find the same problems. Hence, it is sensible to gather resources in a common location. Moreover, most of the CI services provide equivalent environments, while differences are related to the syntax of configuration files.

In the specific case of GitHub Actions, it is not a requirement for each action to be hosted in a separate repository. It is possible to have subdirs, and each of them be a "completely" different action. As a result, it is possible to have one or multiple actions, along with some "generic" scripts, and other specific setup files for "external" CI services (GitLab, Travis, CircleCI, etc.). However, for now I am mostly interested on conciliating and officializing (the best of) existing GitHub Actions.

**Describe how your proposal will work, with code, pseudocode, mockups, and/or diagrams:**

- [113 stars] [godot-ci](https://github.com/marketplace/actions/godot-ci) @aBARICHELLO: "*Docker image to export Godot Engine games. Templates for Gitlab CI/GitHub Actions deploy to GitLab/GitHub Pages or Itch.io*"
  - Despite being published in the marketplace, this does NOT provide any action. On the one hand, a [Dockerfile](https://github.com/aBARICHELLO/godot-ci/blob/master/Dockerfile) is provided (see #1411). On the other hand, an example workflow is provided (see https://github.com/aBARICHELLO/godot-ci/blob/master/.github/workflows/godot-ci.yml). Workflows use GitHub Action's syntax to execute jobs in containers *automagically*.
  - It is mostly a documentation repo, where multiple videotutorials and examples are provided.
  - Deploying to GitHub Pages (from GitLab CI) and to Itch.io (via [Butler](https://itch.io/docs/butler/)) is explained.
  - Another image is provided for projects with Mono enabled.
  - Since the logic is written in bash scripts, the same resources are used in GitHub or GitLab.
- [87 stars] [Godot Export](https://github.com/marketplace/actions/godot-export) @firebelley: "*Automatically exports your Godot games*"
  - It is a JavaScript (TypeScript) GitHub Action that provides an all-in-one solution. It reads the `export_presets.cfg` file, builds all the targets at once and creates releases.
  - Since it needs to download and install godot, the URLs to the tarballs and the templates are required parameters.
  - It supports standard and Mono builds.
- [21 stars] [Build Godot](https://github.com/marketplace/actions/build-godot) @josephbmanley: "*Build a Godot project for multiple platforms*"
  - It is a Docker GitHub Action for exporting project only. No releases, not pages, not itch.io.
  - It based on the Docker image from godot-ci. So, in essence, this repo contains what @aBARICHELLO's repo is lacking, for it to be an Action. From a user's perspective, this is seen in the "complexity" of the workflows:
    - https://github.com/aBARICHELLO/godot-ci/blob/master/.github/workflows/godot-ci.yml#L17-L25
    - https://github.com/josephbmanley/build-godot-action#usage
- [Run GUT Tests](https://github.com/marketplace/actions/run-gut-tests) @josephbmanley: "*Run GUT tests for a Godot project*"
  - Unlike the other action by the same author, this is per se a JavaScript (TypeScript) GitHub Action, which uses `dockerode` (a third-party npm package) to run a container. It feels weird, but the motivation might be the inability for allowing users to customize the Docker image defined in the `action.yml`.
- [godot-tester](https://github.com/marketplace/actions/godot-tester) @croconut: "*Runs all GUT tests you have for your Godot game*"
  - It is a Docker GitHub Action. However, the base Docker image does not include godot. Instead, it is installed in the entrypoint. As a result, it does not really take advantage of using a Docker image.
- [godot-android-builder](https://github.com/marketplace/actions/godot-android-builder) @croconut: "*Builds for Android, Windows, Mac OSX, HTML5 and/or Linux*"
  - Despite the name, it is a Docker GitHub Action that allows to build projects for Android, Windows, MacOS and/or HTML5.
  - Unlike the other action by the same author, this takes advantage of Docker images, and does not install dependencies in the entrypoint. However, the Dockerfile that produces the base image seems not to be publicly available.
  - Compared to godo-ci + build-godot, this allows to build for Android, but it has less options and it is less documented.

NOTE: I am not familiar with testing features. That's why my comments about some of the actions are limited/short.

Overall:

- JavaScript/TypeScript Actions need to be compiled (or `node_modules` needs to be added to the repo). That is really annoying for this kind of projects where any scripting language (bash, Python, GDScript...) works.
- JavaScript/TypeScript Actions can be executed on any of the environments in GitHub Actions (ubuntu, windows or macos). However, Docker Actions (and docker containers) can only be used on ubuntu. As a result:
  - @josephbmanley, I think there is no motivation to write run-gut-tests in JavaScript, because it can only be used on ubuntu anyway. Should it be because you prefer the language, that's ok. However, note that you can achieve the same result with e.g. Python and [docker-py](https://docker-py.readthedocs.io/en/stable/).
    - IMHO, Python is a better fit for this use case. Precisely, run-gut-tests can be useful locally too (for users to run tests without installing godot on their host). I'd expect more people to have a working Python installation, than Node.js.
    - "Locally" means "any laptop/workstation with Docker (Desktop)". Docker is available for Windows and Mac OS, despite it not being available in GitHub Actions (due to issues with nested virtualization).
  - @firebelley, as long as godot-export is supported on any environment, it cannot take advantage of preinstalling dependencies in a container. However:
    - It is sensible to provide a solution that works on any environment, for those use cases that need to be linked/built together with some other platform-specific dependency. Is this something you want to support? Currently, all the examples are on ubuntu, and there is no CI for the action itself.
    - GitHub Actions' caching features might be used for reducing startup time (as an alternative to containers).

My proposal is the following:

- Fork/clone/migrate godot-ci to a new repo named `godotengine/ci` or `godotengine/actions` or `godotengine/ci-helpers` or...
- Merge build-godot into the repo and reorganize the content:
  - A subdir for the Dockerfile(s) of the "base" images.
  - A workflow to generate and publish the "base" images (ref #1411).
  - A subdir for the Docker Action (godot-ci).
  - An example workflow using the syntax from godot-ci.
  - A subdir for docs (GitHub Pages, Itch.io, etc.).
- If godot-export is to be supported on any environment, merge it into the repo, to a subdir.
- run-gut-tests and godot-tester ?
- Enhance the Dockerfile from godot-ci with the features from godot-android-builder. Alternatively, provide another image.

The self-contained approach (publishing the "base" docker image and an action from subdirs of the same repo) is shown in https://github.com/umarcor/hwstudio/tree/godot/root/.github/actions/godot. That is equivalent to the hypothetical first merge of godot-ci and build-godot:

- A workflow is used to build and publish the base image:
  - https://github.com/umarcor/hwstudio/blob/godot/root/.github/workflows/dev.yml
  - https://github.com/umarcor/hwstudio/blob/godot/root/.github/actions/godot/godot.dockerfile (this dockerfile might be located anywhere, it does NOT need to be in the same subdir as the action below)
- A subdir contains a Docker Action, with a custom shell entrypoint:
  - https://github.com/umarcor/hwstudio/blob/godot/root/.github/actions/godot/action.yml
  - https://github.com/umarcor/hwstudio/blob/godot/root/.github/actions/godot/gexport.dockerfile
  - https://github.com/umarcor/hwstudio/blob/godot/root/.github/actions/godot/gexport.sh
  - As commented in #1411, the custom entrypoint should NOT be required, if Godot could find templates in other "standard" locations.
- In another workflow, the Action (and the image) are used/tested:
  - https://github.com/umarcor/hwstudio/blob/godot/root/.github/workflows/build.yml
  - Note that the HTML5 version is uploaded to GitHub Pages. This is equivalent to https://github.com/aBARICHELLO/godot-ci#github-pages. However, it reuses the credentials from actions/checkout.
    - https://github.com/umarcor/hwstudio/blob/godot/root/.github/workflows/build.yml#L31-L33
    - https://github.com/umarcor/hwstudio/blob/godot/root/.github/publish.sh
  - Note also that all artifacts are uploaded (as in godot-ci), but also pushed as the assets of a "nightly" pre-release: https://github.com/umarcor/hwstudio/blob/godot/root/.github/workflows/build.yml#L92-L138 (ref #1412)

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**

I'm quite sure that, the sooner this is conciliated and officialized, the more people will use it. There is not technical limitation. This issue is about communication, documentation, and reducing fragmentation to ease the path for new users.

**Is there a reason why this should be core and not an add-on in the asset library?:**

This question does not apply.

/cc @aBARICHELLO @firebelley @josephbmanley @croconut

**Comment by firebelley** [comment]
Hello! Saw that I was tagged (thank you btw!) and wanted to throw in my 2 cents:
- I would love to see an officially supported action, either primarily as a dockerfile or javascript action. It seems logical that as Godot grows and developers use it in professional/commercial situations, you'd want an easy-to-setup and up-to-date action available that you can just pull in to generate your game builds.
   - In the case of my action, I feel hesitant to develop it further than what's needed for my use-case since my primary goal is to facilitate my own game development. This is where the benefit of an official go-to action comes in. The community can have a single place to look and contribute
- I do, however, understand that having a selection of open-source projects can be a great thing. There's something to be said about having many high-quality community actions available that serve different use cases. The other projects look great, but I think each one has slightly different capabilities or features from the rest.
  - With that said, I'm unsure how useful it would be to consolidate all of these solutions into an official repository. I think that has the potential to be confusing and unmaintainable. But it could be that I'm not seeing all of the benefits of the proposal.

I don't know if those thoughts are helpful at all, but those are the thoughts that come to my mind after reading the proposal. Please correct me if I've misunderstood something!

**Comment by josephbmanley** [comment]
> @josephbmanley, I think there is no motivation to write run-gut-tests in JavaScript, because it can only be used on ubuntu anyway. Should it be because you prefer the language, that's ok. However, note that you can achieve the same result with e.g. Python and docker-py.
The major reason, my newer actions are in Javascript is to reduce build times. When using Docker-based actions, they run a build at the beginning of every run and can end up taking a considerable amount of time. And I've had encounters with people who've requested Windows compatibility for my other actions, so even though no applicable to GUT tests, I don't think it's worth writing them off.

**Comment by umarcor** [comment]
@firebelley, @josephbmanley, I understand your concerns, and I acknowledge that it might be awkward to put those 6 actions together. Some of them have rather specific features and might seems unrelated. Yet, I believe that providing the big picture is important before diving into the details.

> In the case of my action, I feel hesitant to develop it further than what's needed for my use-case since my primary goal is to facilitate my own game development. This is where the benefit of an official go-to action comes in. The community can have a single place to look and contribute

@firebelley, I am not proposing to ditch existing repos/actions. When I use "merge", I mean "pick the interesting features/approaches from that project which are not used in the official solution". And, at the same time, "let existing repos/actions benefit from building on top of an official solution".

Your approach seems to have two distinct and interesting features:

- It could be multi-platform. That's not something you are interested in. It's ok. Nevertheless, if explained, other users/contrbutors might want to pick that task.
- "*automatically reads your export_presets.cfg file to determine which builds to make*" and "*creates a release tagged with a Semantic Version. The zip files containing the export results will be attached to this release*". Hence, from an abstraction point of view, it is a layer on top of other actions (not repos). More on this below.
  - If an official solution allowed to preinstall the dependencies and build exports one by one, your Action might be built on top of that. This would make the action easier to use and to maintain, because `godot_executable_download_url` and `godot_export_templates_download_url` would not be required. It might also reduce startup time. Note that you could do this by setting your JS/TS as the entrypoint of the Docker Action (the complementary scheme of run-gut-tests).

Hence, step by step, my proposal would be to create a new repo containing `godot-ci + build-godot = umarcor/hwstudio/.github/actions`. Then, do a first iteration involving 1) archive build-godot, 2) repurpose godot-ci to contain documentation but no dockerfile, 3) me (and others) start using the official action (and image).

> - I do, however, understand that having a selection of open-source projects can be a great thing. There's something to be said about having many high-quality community actions available that serve different use cases.

I absolutely agree on that. However, specialization and duplication are different issues. It is good to have high-quality specialization. It is very bad to have fragmentation and duplication with no technical benefit. In this specific sense, it is very misleading that:

- godot-ci is announced as an Action; yet, it is not.
- build-godot is the missing piece, but it is located in an unrelated repo. A reference exists, though.
- godot-export seems to do something on top of the others, yet it is completely unrelated.
- None of them can be easily found from the official site, repo or docs.

IMHO, those are the points that need to be addressed first.

> The other projects look great, but I think each one has slightly different capabilities or features from the rest.

While that is true, most of the features are overlapping, and there is no discussion or explanation about what the differences are. From a repository point of view, godot-ci and godot-export are almost equivalent feature-wise. The difference is that godot-ci uses YAML and shell, while godot-export uses JS/TS. There are synctactic differences, but nearly no signifficant feature mismatch.

> With that said, I'm unsure how useful it would be to consolidate all of these solutions into an official repository. I think that has the potential to be confusing and unmaintainable. But it could be that I'm not seeing all of the benefits of the proposal.

I believe it seems confusing and unmaintainable only because multiple languages are used. Regardless, the logic is exactly the same:

1. Install dependencies.
2. Detect the list of "presets" to build, or let the user tell explicitly.
3. Use `godot --export`.
4. Manipulate the packges.
5. Upload artifacts.
6. Create/update GitHub Releases.
7. Update GitHub/GitLab Pages.
8. Publish to Itch.io, Steam, etc.

Hence, you could focus on detecting the presets, manipulating the packages, creating github releases. Those are all currently supported by godot-export, albeit in a monolithic solution. Instead, I believe that a modular solution would achieve better reuse and coordination:

1. Installation is done in a Dockerfile.
2. Merged into 3.
3. Action `godot/action/export`: detect the list of presets and run `godot --export`.
4. Manipulate packages: is custom (`run:`).
5. Upload artifacts: official action `actions/upload-artifacts`.
6. Create/update GitHub Releases: `actions/create-release`, or `eine/tip`, or ...
7. Update GitHub/GitLab Pages: Is there some Action? I use a 20 line script/step (https://github.com/umarcor/hwstudio/blob/godot/root/.github/publish.sh).
8. Publish to Itch.io, Steam, etc.: `godot/action/publish`? `godot/action/publish/itchio`, etc.?

Of course `firebelley/godot-export` might be an specific solution that users can pick as an alternative to 4 and 6. 1-3 would be covered by the "official" godot actions, and I assume that users of godot-export are not interested in 5, 7 and/or 8.

> I don't know if those thoughts are helpful at all, but those are the thoughts that come to my mind after reading the proposal. Please correct me if I've misunderstood something!

I believe they are! Thanks!

Please, do note that I don't expect you to actually implement any of these changes if you don't have the motivation/time to do so. However, I'd like to know whether it makes sense to you.

> The major reason, my newer actions are in Javascript is to reduce build times. When using Docker-based actions, they run a build at the beginning of every run and can end up taking a considerable amount of time.

@josephbmanley, yes, Docker based actions should only add files from the repo to the image. Any installation should be done before. However, if done properly, I think that Docker Actions can be faster. E.g.:

- Docker Action with ready-to-use image: 30s https://github.com/umarcor/hwstudio/actions/runs/223019994
- Docker Action, installing godot and templates in the action image: 60-110s https://github.com/umarcor/hwstudio/actions/runs/222823532

NOTE: it takes longer to updload the artifacts than it takes build the game/tool.

The ready-to-use image takes 522MB (https://hub.docker.com/r/hwstudio/godot/tags), so it is barely larger than retrieving godot and templates at runtime. Docker has some additional benefits:

- The compression is quite optimized.
- The image should be cached/shared between different jobs.

Hence, I'd be curious to know if a JavaScript/TypeScript action can really setup godot (from tarballs) faster.

BTW, it is possible to wrap python/shell entrypoints as JavaScript Actions. This is because python/bash are available in any of the environments. Hence, it is not a requirement to write actions in JS/TS for avoiding the Docker overhead.

> And I've had encounters with people who've requested Windows compatibility for my other actions, so even though no applicable to GUT tests, I don't think it's worth writing them off.

Yes, I agree that it is worth allowing jobs on windows (and probably macos) environments too. However, currently none of the existing solutions can be used on Windows (neither Docker Actions, nor JavaScript/TypeScript Actions). Hence, I would say it seems not to be a priority for the community.
